craft_parts/plugins/base.py

```diff
@@ -122,3 +128,223 @@ def _get_java_post_build_commands(self) -> list[str]:
+        return getattr(self._options, f"{self._options.plugin}_use_uv", False)
+
+    @override
+    def get_build_snaps(self) -> set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        if self._use_uv:
+            return {"astral-uv/latest/beta"}
+        return set()
+
+    @override
+    def get_pull_commands(self) -> list[str]:
+        commands = super().get_pull_commands()
+        if self._use_uv:
+            commands.append("snap alias astral-uv.uv uv || true")
+        return commands
+
+    @override
+    def get_build_packages(self) -> set[str]:
+        """Return a set of required packages to install in the build environment.
+
+        Child classes that need to override this should extend the returned set.
+        """
+        packages = {"findutils", "python3-dev"}
+        if not self._use_uv:
+            packages.add("python3-venv")
+        return packages
+
+    @override
+    def get_build_environment(self) -> dict[str, str]:
+        """Return a dictionary with the environment to use in the build step.
+
+        Child classes that need to override this should extend the dictionary returned
+        by this class.
+        """
+        python3_path = shutil.which("python3")
+        if python3_path is None:
+            raise errors.PluginEnvironmentValidationError(
+                part_name=self._part_info.part_name,
+                reason="cannot find a python3 executable on the system",
+            )
+        return {
+            # Add PATH to the python interpreter we always intend to use with
+            # this plugin. It can be user overridden, but that is an explicit
+            # choice made by a user.
+            "PATH": f"{self._part_info.part_install_dir}/bin:${{PATH}}",
+            "PARTS_PYTHON_INTERPRETER": "python3",
+            "PARTS_PYTHON_VENV_ARGS": "",
+        }
+
+    def _get_venv_directory(self) -> pathlib.Path:
+        """Get the directory into which the virtualenv should be placed."""
+        return self._part_info.part_install_dir
+
+    def _get_create_venv_commands(self) -> list[str]:
+        """Get the commands for setting up the virtual environment."""
+        venv_dir = self._get_venv_directory()
+        venv_commands = (
+            [
+                f'uv venv ${{PARTS_PYTHON_VENV_ARGS}} "{venv_dir}"',
+                f'export UV_PYTHON="{venv_dir}/bin/${{PARTS_PYTHON_INTERPRETER}}"',
+            ]
+            if self._use_uv
+            else [
+                f'"${{PARTS_PYTHON_INTERPRETER}}" -m venv ${{PARTS_PYTHON_VENV_ARGS}} "{venv_dir}"'
+            ]
+        )
+        return [
+            *venv_commands,
+            f'PARTS_PYTHON_VENV_INTERP_PATH="{venv_dir}/bin/${{PARTS_PYTHON_INTERPRETER}}"',
+        ]
+
+    def _get_find_python_interpreter_commands(self) -> list[str]:
+        """Get the commands that find a staged Python interpreter.
```

Comment:
Maybe we could have these large helper scripts available as a separate utility to be invoked at build time?